### PR TITLE
Initial draft for DockerStats metrics aggregation

### PR DESCRIPTION
**Pull request by sagar8192**
I have put together a script to mitigate very High DPM from DockerStats collector on seagull slaves. I have only added one metric right now with MAX aggregation. Idea is to buffer stats in the collector itself, aggregate and flush every few seconds. We are only interested in P50, p75, MAX per image. If this looks good, then I will add more aggregations, more metrics and tests.

### CODE REVIEW

File: src/fullerite/collector/docker_stats.go [hunk 1]
```diff
@@ -177,6 +210,58 @@ func (d *DockerStats) getDockerContainerInfo(container *docker.Container) {
...
+		metrics = append(metrics, metric)
+	}
+	d.log.Warn("Going to flush metrics, count :", len(metrics))
+	d.sendMetrics(metrics)
+	d.statsStore.lastFlush = int32(time.Now().Unix())
+}
+
+func (d *DockerStats) bufferMetrics(container *docker.Container, stats *docker.Stats) {
+	d.statsStore.lock.Lock()
+	d.statsStore.lock.Unlock()
```
Comment (gnufied) [inline src/fullerite/collector/docker_stats.go]: Locking/unlocking immediately?

Comment (sagar8192) [inline src/fullerite/collector/docker_stats.go]: Wanted to add defer d.statsStore.lock.Unlock(). Good Catch!

File: src/fullerite/collector/docker_stats.go [hunk 2]
```diff
@@ -177,6 +210,58 @@ func (d *DockerStats) getDockerContainerInfo(container *docker.Container) {
 	}
 }
 
+func (d *DockerStats) flushMetricsIfNecessary() {
+	if d.aggregateMetrics {
+		if (int32(time.Now().Unix()) - d.statsStore.lastFlush) > 60 {
```
Comment (dichiarafrancesco) [inline src/fullerite/collector/docker_stats.go]: can we put this interval variable instead of 60?

Comment (gnufied) [inline src/fullerite/collector/docker_stats.go]: So, I do not think there is any need to have a flushing interval configured here. If you want metrics to be emitted every 60second, then collector can be configured to run every 60s - it is already possible in fullerite without this change.

**Comment by dichiarafrancesco** [comment]
hey @sagar8192, the code looks pretty much good but i have one concern about this change. the problem is that storing a buffer can be dangerous in terms of memory allocated. so before going over and push this out can you please make some tests to understand the risks behind this change plz?

File: src/fullerite/collector/docker_stats.go [hunk 3]
```diff
@@ -177,6 +210,58 @@ func (d *DockerStats) getDockerContainerInfo(container *docker.Container) {
...
+	rawstats := rawStats{
+		Container: container,
+		Stats:     stats,
+	}
+
+	stringList := strings.Split(container.Config.Image, ":")
+	imageName := stringList[0]
+	if stats, ok := d.statsStore.statsPerImage[imageName]; ok {
+		stats = append(stats, &rawstats)
+	} else {
```
Comment (gnufied) [inline src/fullerite/collector/docker_stats.go]: The way, I am thinking this aggregation should work is - it should work on per-collection basis and no data should be stored after collection has run. This keeps Fullerite pretty lightweight on memory.

So, What we should do is - on each collection, if we only care about lets say `DockerMemoryUsed` for each image, then we should fetch `DockerMemoryUsed` for each container, sum them up and emit with `image_name` as dimension. There is no need to persist any data beyond collection period.

Also, if we don't want certain metrics to be emitted - they can be blacklisted via configuration(Fullerite already supports this).

File: src/fullerite/collector/docker_stats.go [hunk 4]
```diff
@@ -225,6 +310,7 @@ func (d DockerStats) buildMetrics(container *docker.Container, containerStats *d
 
 // sendMetrics writes all the metrics received to the collector channel.
 func (d DockerStats) sendMetrics(metrics []metric.Metric) {
+	d.log.Warn("Sending metrics!!!")
```
Comment (dichiarafrancesco) [inline src/fullerite/collector/docker_stats.go]: do we need this?

**Pull request closed without merging**